akka-stream/src/main/scala/akka/stream/impl/io/SslTls.scala

```diff
@@ -0,0 +1,419 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.nio.ByteBuffer
+import java.security.Principal
+import java.security.cert.Certificate
+import javax.net.ssl.SSLEngineResult.HandshakeStatus
+import javax.net.ssl.SSLEngineResult.HandshakeStatus._
+import javax.net.ssl.SSLEngineResult.Status._
+import javax.net.ssl._
+import akka.actor.{ Props, Actor, ActorLogging, ActorRef }
+import akka.stream.ActorFlowMaterializerSettings
+import akka.stream.impl.FanIn.InputBunch
+import akka.stream.impl.FanOut.OutputBunch
+import akka.stream.impl._
+import akka.util.ByteString
+import akka.util.ByteStringBuilder
+import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+import scala.annotation.tailrec
+import scala.collection.immutable
+import akka.stream.ssl._
+import akka.event.LoggingReceive
+
+object SslTlsCipherActor {
+
+  def props(settings: ActorFlowMaterializerSettings,
+            sslContext: SSLContext,
+            firstSession: NegotiateNewSession,
+            tracing: Boolean,
+            role: StreamTls.Role,
+            closing: StreamTls.Closing): Props =
+    Props(new SslTlsCipherActor(settings, sslContext, firstSession, tracing, role, closing))
+
+  final val TransportIn = 0
+  final val TransportOut = 0
+
+  final val UserOut = 1
+  final val UserIn = 1
+}
+
+class SslTlsCipherActor(settings: ActorFlowMaterializerSettings, sslContext: SSLContext,
+                        firstSession: NegotiateNewSession, tracing: Boolean,
+                        role: StreamTls.Role, closing: StreamTls.Closing)
+  extends Actor with ActorLogging with Pump {
+
+  import SslTlsCipherActor._
+
+  protected val outputBunch = new OutputBunch(outputCount = 2, self, this)
+  outputBunch.markAllOutputs()
+
+  protected val inputBunch = new InputBunch(inputCount = 2, settings.maxInputBufferSize, this) {
+    override def onError(input: Int, e: Throwable): Unit = fail(e)
+  }
+
+  class ChoppingBlock(idx: Int, name: String) extends TransferState {
+    override def isReady: Boolean = buffer.nonEmpty
+    override def isCompleted: Boolean = false
```

Comment:
I thought about this and the reason is that it is very rarely the right thing to just stop this pump (which is what completion would do); in almost all cases we need to tell the engine and keep exchanging data.